### PR TITLE
Rebuild Credential Handling

### PR DESCRIPTION
**Pull request by erikerikson**
## What did you implement:

***Implementing*** [A Potential Fix for] ***Issue***: #1736 

## How did you implement it:

(scoped to AWS)

I migrated a lot of the pre-1.x code forward into 1.x, removing legacy cruft.

Previously you had a number of options, including legacy options for loading credentials.  Given the 0.x=>1.x change, we can drop a lot of the old approaches.  This PR attempts to bring all the good things.

The options for loading credentials are as follows:
1. define credentials on serverless.yml=>service.provider.credentials = { accessKeyId: 'accessKeyId', secretAccessKey: 'secretAccessKey', sessionToken: 'sessionToken' }
2. define a profile from which to get credentials on serverless.yml=>service.provider.profile = 'profile-name' (all profiles loaded using AWS.SharedIniFileCredentials, see http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/SharedIniFileCredentials.html)
3. define credentials for all stages using the standard AWS environment variables (see http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/EnvironmentCredentials.html)
4. define a profile for all stages using the environment variable AWS_PROFILE
5. define credentials for each stage using the standard AWS environment variables with the STAGE name inserted (e.g. stage='test', envVarName='AWS_TEST_*')
6. define a profile for each stage using an environment variable `AWS_${stageName.toUpperCase()}_PROFILE`

If credentials/profiles are declared in multiple ways, the later cases will override the former.

These use cases previously covered all user requirements but the current implemenation allows for an expansion of mechanisms if more mechanisms are desirable.

## How can we verify it:

See the tests that were changed and/or created.  That said, the credential over-ridding logic (see docs) is not explicitly tested.

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config/commands/resources
- [x] Change ready for review message below

***Is this ready for review?:*** Yes

### CODE REVIEW

**Comment by erikerikson** [comment]
This is ready for review

**Comment by rowanu** [comment]
I can see you've got reference to the `AWS_STAGE_PROFILE` variable (and the corresponding secret access key, etc) but support for this (strange, non-standard) profile specification was removed in #2150 by @flomotlik. Can that be removed? I think your fork must be a bit behind `master`...

**Comment by erikerikson** [comment]
Hi @rowanu.  My fork is fully up to date with serverless/master.

The ability to specify different credentials per stage was a strongly demanded feature pre-1.x.  For my group it is a requirement (although we will implement this via serverless.yml, in my opinion).  I took a look at @flomotlik's PR which is concerned that credentials would be set despite the lack of credentials in them.  He even wrote a unit test against that aspect and I made sure this PR passed that test.

Maybe you could share what negative impact, precedent, consequence, or code behavior you are concerned with?

Thanks

**Comment by erikerikson** [comment]
Plenty happy to read up if there's a thread discussing the matter.

**Comment by erikerikson** [comment]
@rowanu, I've checked out your #1940 and #2015.  The behavior you expressed desire for there (stage specific settings should override provider level settings) is enabled by this PR.  The behavior was, in #2015 that only the stage version of the environment variable would be used (since stage was always set) but this PR checks for the most custom and specific set of valid credentials that is specified and uses them.

It doesn't implement the --profile flag and the use of that but it would be easy to add that.  It would fit in nicely as the final source of credentials that is added and checked.

**Comment by rowanu** [comment]
The `AWS_STAGENAME_PROFILE`-style environment variables are unusual and non-standard - they mimic the AWS SDK variables but are not supported by the AWS SDK. I was highlighting that @flomotlik specifically removed it in that PR, so it probably shouldn't be re-added.

You can totally support a per-stage profile using `provider` and `serverless.yml` right now (i.e. after #2015 was merged) - it's what I'm doing right now, and you can see an example [in the docs](https://github.com/serverless/serverless/blob/master/docs/02-providers/aws/01-setup.md#per-stage-profiles).

**Comment by erikerikson** [comment]
Thanks for the clarification @rowanu.  You are right that It is possible to support it on a per-stage basis using the provider in serverless.yml.  That's exactly how I and my group intend to do so.  Actually, I am using the aws-sdk to load the "non-standard" environment variables @ [lib/plugins/aws/index.js:42](https://github.com/serverless/serverless/pull/2229/files#diff-9e9c105390419febe0134cdad91ef22cR42) using the prefix provided form of the environment credentials object.

The environment variable approach mostly serves a sector of our community that uses them on their build servers.  I believe they were not interested in using the credentials file and are additionally in using temporary credentials.  This capability allows Serverless to fit into their existing workflows, including those already built around 0.x.  I'm not personally wedded to the approach since I don't use environment variables myself but I'd be remiss to disable a use case for a sizable portion of the community that hasn't historically been a source of bugs or complaints.

A lot of history on this can be reviewed on PR [#740](https://github.com/serverless/serverless/pull/740) which got a more legacy filled but similar change into v0.5.  Note that there a quite a number of references and links off of that.  Many of them, particularly closed PRs lead to a whole lot of discussion of the matter.  A particularly pertinent discussion is in [#488](https://github.com/serverless/serverless/issues/488).  A larger proportion of the discussion happened in gitter.

**Comment by andymac4182** [comment]
Should `profile` be moved under `credentials` so that they can be loaded via the variable system? Personally a profile is part of the credential loading.

On my local machine I would like to use profiles but as part of our CI solution we will be using Access and Secret Keys.

**Comment by erikerikson** [comment]
Heyya @andymac4182.  I'm not really sure what you're trying to suggest re: "loaded via the variable system".  This PR uses the aws-sdk to load the credentials under the hood.  Can you help me understand?

Personally, I tend to prefer less nesting when sensical.  Maybe more importantly, the profile is simply a named reference to credentials rather than credentials in and of itself.  For those reasons, the current impl (which I didn't change) seemed correct.

That said, I really don't have a strong opinion and would be happy to change it if the community has a strong opinion on the matter.

**Comment by andymac4182** [comment]
The first place it checks is for access key and secret key in a credentials object. I would like to be able to say when access key or secret key are specified as an option then use those if not then use the profile that was specified. I may be missing how this is possible in the current system.

Eg
`sls deploy` uses default profile from variables
`sls deploy --stage {DevName}` uses the devs profile specified in the variables. https://github.com/andymac4182/serverless_example/blob/master/src/serverless.vars.yml#L9 as an example
`sls deploy --stage prod --accesskey XXXXXXX --secretkey YYYYYYY`  uses the access key and secret key not the profile. 

This is to allow for local dev simply where profiles work well and specifying access and secret keys in my CD pipeline where there are many projects being deployed at once so I can't use profiles.

**Comment by erikerikson** [comment]
You could accomplish your use case by using variables in the serverless.yml, including a default stage.  See docs from this commit here:
https://github.com/serverless/serverless/pull/2229/commits/2cfd611329dc031e9156ca98ad2dacfce4b828d2#diff-1456fe13f555062177eeab8824ee9889R116

TL;DR and detailed description of PR creds loading behavior at tail of message.

Note that command line options were not implemented in this PR so that wouldn't be an option.  Instead, environment variables have been used.  It should be straightforward to add CLI params for credentials/profile to this implementation.  I would see the CLI params as the highest-priority source of credentials and would expect that both service-wide and stage-specific values to be supported with stage-specific values overriding service-wide ones.

Traditionally, credentials obtained from a specified profile have overridden any concrete credentials that were provided.  This is currently hard-coded, so long as the two are defined at the same "immediacy level" (e.g. "serverless.yml" < "env vars" < "stage-specific env vars").

It would complicate the code a little but it would be possible to reverse that preference given the user supplies an option to do so.  We could reverse the default too but that would change behavior expectation for existing users.  The traditional rationale, so far as I understand it, was that setting up a profile was a more in depth process, that those credentials were often obtained and refreshed actively by scripts or daemons, and thereby the credentials there were of higher priority.

TL;DR: use profile over credentials, check for these both in each of serverless.yml, env vars, and stage-specific env vars.

A description of the credential loading flow from the commit message (subset):
```
The options for loading credentials are as follows:
1. define credentials on serverless.yml=>service.provider.credentials = { accessKeyId: 'accessKeyId', secretAccessKey: 'secretAccessKey', sessionToken: 'sessionToken' }
2. define a profile from which to get credentials on serverless.yml=>service.provider.profile = 'profile-name' (all profiles loaded using AWS.SharedIniFileCredentials, see http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/SharedIniFileCredentials.html)
3. define credentials for all stages using the standard AWS environment variables (see http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/EnvironmentCredentials.html)
4. define a profile for all stages using the environment variable AWS_PROFILE
5. define credentials for each stage using the standard AWS environment variables with the STAGE name inserted (e.g. stage='test', envVarName='AWS_TEST_*')
6. define a profile for each stage using an environment variable `AWS_${stageName.toUpperCase()}_PROFILE`

If credentials/profiles are declared in multiple ways, the later cases will override the former.
```

**Comment by eahefnawy** [comment]
@erikerikson we're currently discussing this change internally, but this PR is good to go from my side. Will let you know of any updates

Thanks @erikerikson as usual ☺️

**Comment by flomotlik** [comment]
@erikerikson great job. I like the change and the addition of support for Stage scoped environment variables. I didn't like the implementation before and removed it because it was unclear in general and added too much confusion imho. This is resolved here as it has much cleaner code and documentation.

I've only looked over the docs yet and haven't played with it, which I will do tomorrow to test it out. After that this is imho good to go.

File: docs/02-providers/aws/01-setup.md [hunk 1]
```diff
@@ -69,37 +69,132 @@ Default output format [None]: ENTER
...
+```
+```bash
+export AWS_PROD_ACCESS_KEY_ID=<prod-key>
+export AWS_PROD_SECRET_ACCESS_KEY=<prod-secret>
+export AWS_PROD_SESSION_TOKEN=<prod-token>
+
+serverless <...>
 ```
 
+Et cetera
```
Comment (flomotlik) [inline docs/02-providers/aws/01-setup.md]: Can you remove this and the next line

File: docs/02-providers/aws/01-setup.md [hunk 2]
```diff
@@ -69,37 +69,132 @@ Default output format [None]: ENTER
...
-As an advanced use-case, you can deploy different stages to different accounts by using different profiles per stage. In order to use different profiles per stage, you must leverage [variables](../../01-guide/08-serverless-variables.md) and the provider profile setting.
+One profile for all stages using serverless.yml
+```yml
+provider:
+  profile: your-profile
+```
+
+A profile for each stage using serverless.yml
+```yml
+vars:
```
Comment (flomotlik) [inline docs/02-providers/aws/01-setup.md]: Currently vars doesn't work, it needs to be in ***custom***. Could you change all occurences of vars to custom for now (I think we should support vars in the future though

Comment (erikerikson) [inline docs/02-providers/aws/01-setup.md]: Saw this but also that the branch was merged.  I can implement these changes but based on [this code](https://github.com/serverless/serverless/blob/master/lib/classes/Variables.js#L140), it seems this should work.  Adding the following to the tests [here](https://github.com/serverless/serverless/blob/master/tests/classes/Variables.js#L338) succeeds:

```
  describe('#getValueFromSelf()', () => {
    it('should get variable from self serverless.yml file', () => {
      const serverless = new Serverless();
      serverless.variables.service = {
        someVar: 'someValue',
        service: 'testService',
        provider: 'testProvider',
        defaults: serverless.service.defaults,
      };

      serverless.variables.loadVariableSyntax();

      const valueToPopulate = serverless.variables.getValueFromSelf('self:someVar');
      expect(valueToPopulate).to.be.equal('someValue');
    });
  });
```

Just making sure that we're good on this

Comment (erikerikson) [inline docs/02-providers/aws/01-setup.md]: don't get me wrong, happy to change naming to comply with standard variable naming in docs, just want to make the change for the right reasons

File: docs/02-providers/aws/01-setup.md (hunk 1 above)
Comment (erikerikson) [inline docs/02-providers/aws/01-setup.md]: this and prior implemented here: https://github.com/serverless/serverless/pull/2297
